staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -228,80 +228,17 @@ func (dm *discoveryManager) fetchFreshDiscoveryForService(gv metav1.GroupVersion
 
 		dm.setCacheEntryForService(info.service, cached)
 		return &cached, nil
-	case http.StatusNotAcceptable:
-		// Discovery Document is not being served at all.
-		// Fall back to legacy discovery information
-		if len(gv.Version) == 0 {
-			return nil, errors.New("not found")
-		}
-
-		var path string
-		if len(gv.Group) == 0 {
-			path = "/api/" + gv.Version
-		} else {
-			path = "/apis/" + gv.Group + "/" + gv.Version
-		}
-
-		req, err := http.NewRequest("GET", path, nil)
-		if err != nil {
-			// NewRequest should not fail, but if it does for some reason,
-			// log it and continue
-			return nil, fmt.Errorf("failed to create http.Request: %v", err)
-		}
-
-		// Apply aggregator user to request
-		req = req.WithContext(
-			request.WithUser(
-				req.Context(), &user.DefaultInfo{Name: "system:kube-aggregator"}))
-
-		// req.Header.Add("Accept", runtime.ContentTypeProtobuf)
-		req.Header.Add("Accept", runtime.ContentTypeJSON)
-
-		if exists && len(cached.etag) > 0 {
-			req.Header.Add("If-None-Match", cached.etag)
-		}
-
-		writer := newInMemoryResponseWriter()
-		handler.ServeHTTP(writer, req)
-
-		if writer.respCode != http.StatusOK {
-			return nil, fmt.Errorf("failed to download discovery for %s: %v", path, writer.String())
-		}
-
-		parsed := &metav1.APIResourceList{}
-		if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), writer.data, parsed); err != nil {
-			return nil, err
-		}
-
-		// Create a discomap with single group-version
-		resources, err := endpoints.ConvertGroupVersionIntoToDiscovery(parsed.APIResources)
-		if err != nil {
-			return nil, err
-		}
-
-		discoMap := map[metav1.GroupVersion]apidiscoveryv2beta1.APIVersionDiscovery{
-			// Convert old-style APIGroupList to new information
-			gv: {
-				Version:   gv.Version,
-				Resources: resources,
-			},
-		}
-
-		cached = cachedResult{
-			discovery:   discoMap,
-			lastUpdated: now,
-		}
-
-		// Do not save the resolve as the legacy fallback only fetches
-		// one group version and an API Service may serve multiple
-		// group versions.
-		return &cached, nil
-
 	case http.StatusOK:
 		parsed := &apidiscoveryv2beta1.APIGroupDiscoveryList{}
 		if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), writer.data, parsed); err != nil {
-			return nil, err
+			klog.V(3).Infof("DiscoveryManager: Failed to decode response for %s: %v", info.service.String(), err)
+			break
 		}
+		if parsed.Kind != "APIGroupDiscoveryList" || parsed.APIVersion != "v2beta1" {
```

Comment:
I noticed `runtime.DecodeInto` doesn't fail even the Kind is incorrect. Open to other suggestions of checking deserialization failures since I feel like there'd be a simpler way.